hubtools/src/lib.rs

```diff
@@ -652,23 +716,67 @@ impl RawHubrisArchive {
 
         let board = manifest
             .as_table()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .get("board")
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .as_str()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .to_owned();
         let mut name = manifest
             .as_table()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .get("name")
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .as_str()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .to_owned();
 
+        // The optional epoch argument and optional app.toml epoch
+        // parameter both default to zero and must agree with each other.
+        let epoch_arg: u32 = if let Some(arg) = epoch {
+            arg.parse::<u32>()
+                .map_err(|_| Error::InvalidEpoch(arg.to_string()))?
+        } else {
+            0u32
+        };
+
+        let epoch_toml: u32 = manifest
+            .as_table()
+            .ok_or(Error::BadTomlTypeEpoch)?
+            .get("epoch")
+            .ok_or(Error::BadTomlTypeEpoch)?
+            .as_integer()
+            .ok_or(Error::InvalidEpoch(format!("{}:{}", file!(), line!())))?
+            .try_into()
+            .map_err(|_| {
+                Error::InvalidEpoch(format!("{}:{}", file!(), line!()))
+            })?;
+
+        // Default to zero or the only non-zero value,
+        // otherwise, they need to match.
+        let epoc = match (epoch_arg, epoch_toml) {
+            (0, 0) => Ok(0),
+            (x, 0) => Ok(x),
+            (0, y) => Ok(y),
+            (x, y) => {
+                if x == y {
+                    Ok(x)
+                } else {
+                    Err(Error::InvalidEpoch(format!(
+                        "epoch value from app.toml({}) != arg({})",
```

Comment:
Why do we need to provide the epoch as an argument at all, if it must match the one from `app.toml`?  Should `app.toml` be the canonical location of the epoch value?

Reply:
I'm taking epoch and version out of all of the Hubris toml files as well as the fake toml we make for Bootleby. It's going to be injected according to a `sprot-release` action. See `sprot-release` `docs/howtorelease.md` which will be updated to handle the `epoch` value.